libraries/AP_ExternalAHRS/AP_ExternalAHRS_GSOF.cpp

```diff
@@ -0,0 +1,362 @@
+
+AP_ExternalAHRS_GSOF::AP_ExternalAHRS_GSOF(AP_ExternalAHRS *_frontend,
+        AP_ExternalAHRS::state_t &_state): AP_ExternalAHRS_backend(_frontend, _state)
+{
+
+    auto &sm = AP::serialmanager();
+    uart = sm.find_serial(AP_SerialManager::SerialProtocol_AHRS, 0);
+    if (!uart) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, LOG_FMT, get_name(), "no UART");
+        return;
+    }
+    baudrate = sm.find_baudrate(AP_SerialManager::SerialProtocol_AHRS, 0);
+    port_num = sm.find_portnum(AP_SerialManager::SerialProtocol_AHRS, 0);
+    pktbuf = NEW_NOTHROW uint8_t[AP_GSOF::MAX_PACKET_SIZE];
+
+    if (!pktbuf) {
+        AP_BoardConfig::allocation_error("GSOF ExternalAHRS");
+    }
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_ExternalAHRS_GSOF::update_thread, void), "AHRS", 2048, AP_HAL::Scheduler::PRIORITY_SPI, 0)) {
+        AP_BoardConfig::allocation_error("GSOF ExternalAHRS failed to allocate ExternalAHRS update thread");
+    }
+
+    // Offer GPS even through it's a tightly coupled EKF.
+    set_default_sensors(uint16_t(AP_ExternalAHRS::AvailableSensor::GPS));
+
+    hal.scheduler->delay(5000);
+    if (!initialised()) {
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING, LOG_FMT, get_name(), "missing data within 5s.");
+    }
+}
+
+// get serial port number for the uart
+int8_t AP_ExternalAHRS_GSOF::get_port(void) const
+{
+    if (!uart) {
+        return -1;
+    }
+    return port_num;
+};
+
+void AP_ExternalAHRS_GSOF::update_thread(void)
+{
+    // TODO configure receiver to output expected data.
+
+    auto last_debug = AP_HAL::millis();
+    size_t pps = 0;
+
+    uart->begin(baudrate);
+    
+    while (true) {
+        // TODO should we ever call begin() again?
+        while (uart->available() > 0) {
+            const uint8_t c = uart->read();
+
+            AP_GSOF::MsgTypes parsed;
+            const auto parse_res = parse(c, parsed);
+            if (parse_res != PARSED_GSOF_DATA) {
+                continue;
+            }
+
+            auto const now = AP_HAL::millis();
+            pps++;
+
+            if (now - last_debug > 1000) {
+                GCS_SEND_TEXT(MAV_SEVERITY_INFO,
+                    "GSOF PPS: %lu, ins %u",
+                    static_cast<unsigned long>(pps),
+                    static_cast<uint16_t>(ins_full_nav.gnss_status));
+                last_debug = now;
+                pps = 0;
+            }
+            if (parsed.get(AP_GSOF::POS_TIME)) {
+                last_pos_time_ms = now;
+
+                gps_data.satellites_in_view = pos_time.num_sats;
+            }
+
+            if (parsed.get(AP_GSOF::INS_FULL_NAV)) {
+                last_ins_full_nav_ms = now;
+
+                gps_data.gps_week = ins_full_nav.gps_week;
+                gps_data.ms_tow = ins_full_nav.gps_time_ms;
+                gps_data.ned_vel_north = ins_full_nav.vel_n;
+                gps_data.ned_vel_east = ins_full_nav.vel_e;
+                gps_data.ned_vel_down = ins_full_nav.vel_d;
```

Comment:
This doesn't look right - this isn't gps data, this is a position estimate based on sensor fusion!

So this data is what you should be returning as an AHRS backend, and `LLH_MSL` presumably is raw sensor data from the GPS attached to the Trimble so *should* be used to fill gps_data


... and if you want to argue the point that should be in the form of comments, because this is decidedly odd :-)